plugins/in_calyptia_fleet/in_calyptia_fleet.c

```
@@ -228,11 +216,165 @@ flb_sds_t fleet_config_filename(struct flb_in_calyptia_fleet_config *ctx, char *
+            flb_plg_error(ctx->ins, "fleet_config_deref: line from reference file was truncated: %s", ref_filename);
+            fclose(ref_file_ptr);
+            flb_sds_destroy(ref_filename);
+            return NULL;
+        }
+
+        /* Remove trailing newline if present */
+        if (len > 0 && line[len - 1] == '\n') {
+            line[len - 1] = '\0';
+        }
+        config_path = flb_sds_create(line);
+    }
+
+    fclose(ref_file_ptr);
+    flb_sds_destroy(ref_filename);
+
+    return config_path;
+}
+
+/**
+ * Update ref_name's ref file to contain config_path.
+ * If the ref file does not exist it is created, otherwise its contents are
+ * overwritten. This returns FLB_TRUE if successful, FLB_FALSE otherwise.
+ */
+static int fleet_config_set_ref(struct flb_in_calyptia_fleet_config *ctx,
+                                const char *ref_name,
+                                const char *config_path)
+{
+    flb_sds_t ref_filename;
+    flb_sds_t temp_filename;
+    FILE *temp_file_ptr;
+    size_t to_write;
+    size_t written;
+
+    if (ctx == NULL) {
+        return FLB_FALSE;
+    }
+    if (ref_name == NULL || config_path == NULL) {
+        flb_plg_error(ctx->ins, "fleet_config_set_ref: ref_name or config_path is NULL");
+        return FLB_FALSE;
+    }
+
+    ref_filename = fleet_config_ref_filename(ctx, ref_name);
+    if (ref_filename == NULL) {
+        flb_plg_error(ctx->ins, "fleet_config_set_ref: unable to get ref filename for %s", ref_name);
+        return FLB_FALSE;
+    }
+
+    /* First write to a temporary file in case the write wholly or partially fails */
+    temp_filename = flb_sds_create_size(flb_sds_len(ref_filename) + 5);
+    if (temp_filename == NULL) {
+        flb_plg_error(ctx->ins, "fleet_config_set_ref: unable to create temp filename");
+        flb_sds_destroy(ref_filename);
+        return FLB_FALSE;
+    }
+    temp_filename = flb_sds_printf(&temp_filename, "%s.tmp", ref_filename);
```

Should we check if `temp_filename` == NULL again?

---

Yeah, updated. At least we don't need another temp var since `flb_sds_printf` updates the input pointer if the underlying buffer is reallocated.

---

There are so many other unrelated examples of `flb_sds_printf` errors being ignored in this file but I'd prefer to update those separately since this diff is already big.